### DISCUSSION TITLE
ヨミガナ検索について

### DISCUSSION THREAD

**Discussion by blueowlgreenfish** [pgroonga/pgroonga#243] [Q&A] [answered]
[MySQL・PostgreSQLだけで作る高速あいまい全文検索システム](https://slide.rabbit-shocker.org/authors/kou/db-tech-showcase-tokyo-2018/)のスライドの中の、PGroongaのヨミガナ検索のチュートリアルをしていたのですが、`SELECT name FROM menus WHERE name &@~ '焼きにく';`で検索したら、焼きにく定食しか返ってきません。どうしたら、焼肉定食も返すことができるでしょうか？
ちなみに、`echo 焼肉と焼きにく | mecab | cut -d, -f8`を実行すると、

> ヤキニク
ト
ヤキニク

ではなく、

> ヤキニク
ト
ヤキ
ニク

が返ってきます。これも何か関係があるのでしょうか？
Postgresqlのバージョンは14.5。
PGroongaのバージョンは2.4.0。
MeCabのバージョンは0.996で、IPA辞書のバージョンはmecab-ipadic-2.7.0-20070801です。
よろしくお願いします。

**Comment by komainu8** [comment, marked as answer]
ご報告ありがとうございます！
手元でも検証してみますが、 問題が起きている環境で`EXPLAIN ANALYZE VERBOSE SELECT name FROM menus WHERE name &@~ '焼きにく';`を実行した結果を教えてもらえますか？

**Reply by blueowlgreenfish** [reply]
次の結果が返ってきました。

>  Index Scan using menus_search on public.menus  (cost=0.00..46.80 rows=1 width=32) (actual time=0.949..0.954 rows=1 loops=1)
   Output: name
   Index Cond: (menus.name &@~ '焼きにく'::text)
 Planning Time: 1.070 ms
 Execution Time: 1.201 ms
(5 rows)

**Reply by komainu8** [reply]
ありがとうございます！
こちらでも再現しました。

原因は、ご提示頂いたとおり mecab-ipadic-2.7.0-20070801 では、
"焼肉" -> "ヤキニク"
"焼きにく" -> "ヤキ", "ニク"
と解析されてしまうことでした。

"焼きにく"->"ヤキニク"と解析されれば、"焼肉"の解析結果("焼肉" -> "ヤキニク")と一致するのでヒットしますが、"焼きにく"は、前述の通り "焼きにく" -> "ヤキ", "ニク" と解析されるので、"焼肉"の解析結果と一致せず、 "焼きにく"で検索しても"焼肉"はヒットしません。

解決方法としては、 IPA辞書ではなく、 [NAIST Japanese Dictionary](https://ja.osdn.net/projects/naist-jdic/) をお使いいただくことで"焼きにく"で検索したときに"焼肉"がヒットします。

NAIST Japanese Dictionary に切り替えた後は、 REINDEX でPGroongaのインデックスを再構築してください。
再構築後でもヒットしない場合は、以下を確認してください。

`EXPLAIN ANALYZE VERBOSE SELECT name FROM menus WHERE name &@~ '焼きにく';`の結果が` Seq Scan`となっていると思いますので、その場合は `SET enable_seqscan = off;` を実行してから再度 `SELECT name FROM menus WHERE name &@~ '焼きにく';`を実行してみてください。

**Reply by blueowlgreenfish** [reply]
NAIST Japanese Dictionaryに切り替えたことで、”焼肉定食”と”焼きにく定食”、両方ヒットしました！
ありがとうございました。
